### ISSUE TITLE
我如何基于一个qwen2.5vl创建一个新的reward model结构并进行训练

### ISSUE THREAD

**Issue by Tunanzzz** [modelscope/ms-swift#4635]
我希望基于qwen2.5vl创建一个新的reward model结构，这个reward model相较于qwen2.5vl的差异就是会添加一个全连接层(hidden_size, 1)，将generate的最后一个token的最后一层特征通过全连接到reward分数，我希望添加这个模型并进行训练，我应该怎么做？

**Comment by hjh0119** [comment]
reward modeling
```
swift rlhf \
--rlhf_type rm
```

**Comment by Tunanzzz** [comment]
感谢你的回答，我现在新的问题，请问数据格式应该是什么样的？以及我希望能够拿到模型针对我的question的输出，并对输出做一个类似于dpo的kl损失与原本的reward分类损失加权，应该怎么做

**Comment by Tunanzzz** [comment]
我在ms-swift/swift/trainers/rlhf_trainer/reward_trainer.py中添加了打印信息的部分：
`class RewardTrainer(RLHFTrainerMixin, SwiftMixin, HFRewardTrainer):

    def compute_loss(self,
                     model: Union[PreTrainedModel, nn.Module],
                     inputs: Dict[str, Union[torch.Tensor, Any]],
                     return_outputs=False,
                     num_items_in_batch=None) -> Union[torch.Tensor, Tuple[torch.Tensor, Dict[str, torch.Tensor]]]:
                     
        print("*******", inputs)
        print("`````````--------------------")
        inputs.pop('labels', None)  # not use
        attention_mask = inputs['attention_mask']
        batch_size = attention_mask.shape[0] // 2
        rewards = model(**inputs).logits
        rewards_chosen, rewards_rejected = torch.split(rewards, batch_size, dim=0)
        if 'margin' in inputs:
            loss = -nn.functional.logsigmoid(rewards_chosen - rewards_rejected - inputs['margin']).mean()
        else:
            loss = -nn.functional.logsigmoid(rewards_chosen - rewards_rejected).mean()
        if self.args.center_rewards_coefficient is not None:
            loss += self.args.center_rewards_coefficient * torch.mean((rewards_chosen + rewards_rejected)**2)
        # compat transformers>=4.46.*
        if num_items_in_batch is not None and self.model_accepts_loss_kwargs:
            loss /= self.args.gradient_accumulation_steps
        if return_outputs:
            return loss, {
                'rewards_chosen': rewards_chosen,
                'rewards_rejected': rewards_rejected,
            }
        return loss`


这是我的sh文件：
`nproc_per_node=4
MAX_PIXELS=409600 \
CUDA_VISIBLE_DEVICES=3,4,6,7 \
NPROC_PER_NODE=$nproc_per_node \
swift rlhf \
    --rlhf_type rm \
    --model /home/zhoupc/safe_alignment/models/qwen-2.5-vl-7B \
    --train_type lora \
    --dataset /home/zhoupc/safe_alignment/saferlhf-v/safe_rlhf_v/BeaverTails-V/swift_rm.jsonl \
    --torch_dtype bfloat16 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-4 \
    --lora_rank 8 \
    --freeze_vit true \
    --lora_alpha 32 \
    --gradient_accumulation_steps $(expr 16 / $nproc_per_node) \
    --eval_steps 100 \
    --save_steps 10000 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --max_length 4096 \
    --save_only_model true \
    --output_dir /home/zhoupc/safe_alignment/ms-swift/checkpoint/safe-rlhf-v-rm \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --deepspeed zero3 \
    --dataset_num_proc 4`

我发现训练正常启用的时候，不会输出我写的print的内容，这是因为rlhf_type=rm的时候调用的并不是ms-swift/swift/trainers/rlhf_trainer/reward_trainer.py的compute_loss方法吗？

**Comment by hjh0119** [comment]
Check the path to swift in your environment; it may not be in ms-swift/swift.

```
pip show ms-swift
```

**Comment by Tunanzzz** [comment]
`root@9e02fadf3a84:/home/zhoupc/safe_alignment/ms-swift# pip show ms-swift
Name: ms_swift
Version: 3.4.1.post1
Summary: Swift: Scalable lightWeight Infrastructure for Fine-Tuning
Home-page: https://github.com/modelscope/swift
Author: DAMO ModelScope teams
Author-email: contact@modelscope.cn
License: Apache License 2.0
Location: /usr/local/lib/python3.11/site-packages
Requires: accelerate, addict, aiohttp, attrdict, binpacking, charset-normalizer, cpm-kernels, dacite, datasets, einops, fastapi, gradio, importlib-metadata, jieba, matplotlib, modelscope, nltk, numpy, openai, oss2, pandas, peft, pillow, requests, rouge, safetensors, scipy, sentencepiece, simplejson, sortedcontainers, tensorboard, tiktoken, tqdm, transformers, transformers-stream-generator, trl, uvicorn, zstandard
Required-by:`


我添加print的代码在：/home/zhoupc/safe_alignment/ms-swift/swift/trainers/rlhf_trainer/reward_trainer.py

**Comment by Tunanzzz** [comment]
我知道问题所在了，谢谢你的回复！！！

**Comment by Tunanzzz** [comment]
我想问一下rlhf_type为rm的时候，输入数据的格式是什么样的，有哪些参数可以设置

**Comment by Tunanzzz** [comment]
> Check the path to swift in your environment; it may not be in ms-swift/swift.
> 
> ```
> pip show ms-swift
> ```

感谢你的回复，按照你的提示，我解决了这个问题
